packages/server/src/sse/sse.service.ts

```diff
@@ -1,34 +1,97 @@
-import { Injectable } from '@nestjs/common';
+import { Injectable, OnModuleDestroy } from '@nestjs/common';
+import { each } from 'async';
 import { serialize } from 'class-transformer';
 import * as apm from 'elastic-apm-node';
 import { Response } from 'express';
+import { RedisService } from 'nestjs-redis';
 
 export interface Client<T> {
   metadata: T;
   res: Response;
 }
+
+interface Connection {
+  res: Response;
+  cleanup: () => Promise<void>;
+}
+
+interface RedisClientInfo<T> {
+  clientId: number;
+  metadata: T;
+}
 /**
  * T is metadata associated with each Client
  *
  * Low level abstraction for sending SSE to "rooms" of clients.
  * Probably don't use this directly, and wrap it in a service specific to that event source
+ *
+ * This handles when there's multiple backend instances by assigning unique client ids to each connection.
+ * When one instance wants to send to a client, it publishes to a Redis channel for the client.
+ * All instances listen to Redis, and if they are the one managing that client, they send the msg.
+ *
+ * Rooms with client metadata are also maintained in Redis key/value store.
```

Comment:
This service got a good bit more complex, but read this comment to get the gist of it.